
<h1> Battle of the Neighbourhoods - Toronto </h1>

Author: Rohit Kumar Gupta

This notebook contains Questions 1 and 2 of the Assignment.

In [ ]:
import pandas as pd

# Question 1

## Importing Data

In [ ]:
import requests

In [ ]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means that we are able to make the connection to the page

In [ ]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code  ...                                      Neighbourhood
 0           M1A  ...                                       Not assigned
 1           M2A  ...                                       Not assigned
 2           M3A  ...                                          Parkwoods
 3           M4A  ...                                   Victoria Village
 4           M5A  ...                          Regent Park, Harbourfront
 ..          ...  ...                                                ...
 175         M5Z  ...                                       Not assigned
 176         M6Z  ...                                       Not assigned
 177         M7Z  ...                                       Not assigned
 178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...
 179         M9Z  ...                                       Not assigned
 
 [180 rows x 3 columns],
                                                   0   ...   17
 0                               

In [ ]:
len(wiki_data), type(wiki_data)

(3, list)

We need the first table alone, so dropping the other tables

In [ ]:
wiki_data = wiki_data[0]
wiki_data

Postal Code  ...                                      Neighbourhood
0           M1A  ...                                       Not assigned
1           M2A  ...                                       Not assigned
2           M3A  ...                                          Parkwoods
3           M4A  ...                                   Victoria Village
4           M5A  ...                          Regent Park, Harbourfront
..          ...  ...                                                ...
175         M5Z  ...                                       Not assigned
176         M6Z  ...                                       Not assigned
177         M7Z  ...                                       Not assigned
178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...
179         M9Z  ...                                       Not assigned

[180 rows x 3 columns]

Dropping Borough which are not assigned

In [ ]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code  ...                                      Neighbourhood
2           M3A  ...                                          Parkwoods
3           M4A  ...                                   Victoria Village
4           M5A  ...                          Regent Park, Harbourfront
5           M6A  ...                   Lawrence Manor, Lawrence Heights
6           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
160         M8X  ...      The Kingsway, Montgomery Road, Old Mill North
165         M4Y  ...                               Church and Wellesley
168         M7Y  ...  Business reply mail Processing Centre, South C...
169         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [ ]:
df = df.groupby(['Postal Code']).head()
df

Postal Code  ...                                      Neighbourhood
2           M3A  ...                                          Parkwoods
3           M4A  ...                                   Victoria Village
4           M5A  ...                          Regent Park, Harbourfront
5           M6A  ...                   Lawrence Manor, Lawrence Heights
6           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
160         M8X  ...      The Kingsway, Montgomery Road, Old Mill North
165         M4Y  ...                               Church and Wellesley
168         M7Y  ...  Business reply mail Processing Centre, South C...
169         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [ ]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [ ]:
df = df.reset_index()
df

index  ...                                      Neighbourhood
0        2  ...                                          Parkwoods
1        3  ...                                   Victoria Village
2        4  ...                          Regent Park, Harbourfront
3        5  ...                   Lawrence Manor, Lawrence Heights
4        6  ...        Queen's Park, Ontario Provincial Government
..     ...  ...                                                ...
98     160  ...      The Kingsway, Montgomery Road, Old Mill North
99     165  ...                               Church and Wellesley
100    168  ...  Business reply mail Processing Centre, South C...
101    169  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
102    178  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 4 columns]

In [ ]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code  ...                                      Neighbourhood
0           M3A  ...                                          Parkwoods
1           M4A  ...                                   Victoria Village
2           M5A  ...                          Regent Park, Harbourfront
3           M6A  ...                   Lawrence Manor, Lawrence Heights
4           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
98          M8X  ...      The Kingsway, Montgomery Road, Old Mill North
99          M4Y  ...                               Church and Wellesley
100         M7Y  ...  Business reply mail Processing Centre, South C...
101         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
102         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

In [ ]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

# Question 2

Installing geocoder

In [ ]:
pip install geocoder

     |████████████████████████████████| 102kB 2.2MB/s 


In [ ]:
import geocoder # import geocoder

Tried the below approach, ran for 20 mins, then killed it. Changing the code cell to Text for now so that the run all execution doesn't stop.

```python
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
```

Alternatively, as suggested in the assignment, Importing the CSV file from the URL

In [ ]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [ ]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [ ]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [ ]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [ ]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  ...   Latitude  Longitude
0           M3A        North York  ...  43.753259 -79.329656
1           M4A        North York  ...  43.725882 -79.315572
2           M5A  Downtown Toronto  ...  43.654260 -79.360636
3           M6A        North York  ...  43.718518 -79.464763
4           M7A  Downtown Toronto  ...  43.662301 -79.389494
..          ...               ...  ...        ...        ...
98          M8X         Etobicoke  ...  43.653654 -79.506944
99          M4Y  Downtown Toronto  ...  43.665860 -79.383160
100         M7Y      East Toronto  ...  43.662744 -79.321558
101         M8Y         Etobicoke  ...  43.636258 -79.498509
102         M8Z         Etobicoke  ...  43.628841 -79.520999

[103 rows x 5 columns]

In [ ]:
combined_data.shape

(103, 5)

**Solution:** We get 103 rows as expected when we do a inner join, so we have good data.